### PR TITLE
Add merge request labels to environment variables 

### LINKED ISSUE #900: Accessing labels in build
## Issue

Note: MR #646 would've helped me here, but it was never merged.

### Context
- **Gitlab plugin version**: 1.5.11
- **Gitlab version**: 11.7.5
- **Jenkins version**: 2.89.4
- **Job type**: Pipeline

### Logs & Traces

Output from `echo sh(returnStdout: true, script: 'env')` in build (only showing variables related to this plugin):

```gitlabMergeRequestTargetProjectId=56
gitlabSourceRepoURL=git@git.example.com:group/project.git
gitlabTargetRepoHttpUrl=https://git.example.com/group/project.git
gitlabMergedByUser=anton-johansson
gitlabTargetRepoSshUrl=git@git.example.com:group/project.git
gitlabTargetBranch=master
gitlabMergeRequestIid=1023
gitlabSourceRepoHttpUrl=https://git.example.com/group/project.git
gitlabMergeRequestState=opened
gitlabMergeRequestTitle=Test
gitlabMergeRequestLastCommit=d4f4b002445261f691ec45bffc4dc81534e50f0e
gitlabSourceRepoSshUrl=git@git.example.com:group/project.git
gitlabTargetRepoName=Project
gitlabTargetNamespace=Group
gitlabSourceRepoHomepage=https://git.example.com/group/project
gitlabBranch=anton-test
gitlabSourceBranch=anton-test
gitlabUserEmail=antoon.johansson@gmail.com
gitlabMergeRequestId=5618
gitlabActionType=MERGE
gitlabSourceRepoName=Project
gitlabSourceNamespace=Group
gitlabUserName=Anton Johansson
```

### Problem description

I would like to access MR labels so I can perform various logic depending on them. My main idea here is to automatically deploy to my Kubernetes cluster if a certian label is set. I want this because I don't want to push Docker images to our repository for every commit in every MR. Only the ones that are relevant to deploy.

This also means that I want to be able to trigger a build by only adding a label (this specific label, if possible). Issue #871 states that this is already happening, but it isn't for me.

### PR DESCRIPTION
**Pull request by yiftahw**
**many thanks to @anton-johansson at #901 and @jakub-bochenski at #646 for starting this work**  
closes #900 

### Done
- rebased work from #901 to latest master branch
- provided test cases at `GitLabEnvironmentContributorTest.java`
- added pipeline helper function `GitLabMergeRequestLabelExists("some label") -> Boolean`
- added explanation to `README.md`
- provided test cases at `GitLabMergeRequestLabelExistsStepTest.java`
- indentation fix in `src/docker/README.md`

#### groovy example of using the helper function:
```groovy
if (GitLabMergeRequestLabelExists("test label")) {
    echo "test label found"
}
```
#### groovy example of echoing all the labels:
```groovy
// Check if the environment variable is null (should never be empty - see unit tests)
if (env.gitlabMergeRequestLabels == null) {
    echo "No merge request labels provided."
} else {
    // Split the labels by commas and trim each label
    def labels = env.gitlabMergeRequestLabels.split(',').collect { it.trim() }
    echo "Merge request labels:"
    labels.each { label -> echo label }
}
```

### Testing done
  - test set up:
    - OpenSUSE Leap 15.6 VM
    - Java 17 (openjdk)
    - Maven 3.9.9
    - docker running inside the VM
      - **changed the `Jenkins` tag to `2.479` (with `lts` I got an error when trying to install the plugin: Jenkins minimum version required: 2.479)**
      - GitLab tags tested: `latest` (`17.5`), `15.11.13-ce.0`
- built the plugin with `mvn deploy` and manually installed to Jenkins with the web UI.
- created a pipeline in `Jenkins`, integrated it to `GitLab` and observed the behavior of new environment variable and helper function in push and merge request hooks.  
  
### Note
- @anton-johansson provided a note in his merge request #901 that this doesn't work when putting comments to merge requests, because the webhook doesn't contain the labels.
- using the GitLab Jenkins Integration (under `Settings -> Integrations -> Jenkins`) doesn't send webhooks on comments.
- using a  generic web trigger  (under `Settings -> Webhooks`) allows sending hooks on comments, and according to the oldest doc online - [GitLab 15.6 Webhook events](https://archives.docs.gitlab.com/15.6/ee/user/project/integrations/webhook_events.html#comment-on-a-merge-request) it does send the labels on comments in merge requests.
- the class `NoteHook` doesn't handle the labels

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/test/java/com/dabsquared/gitlabjenkins/environment/GitLabEnvironmentContributorTest.java [hunk 1]
```diff
@@ -74,46 +74,43 @@ public void matrixProjectTest() throws IOException, InterruptedException, Execut
         }
     }
 
-    public void testFreeStyleProjectNoLabelsBase(CauseData causeData)
+    public void testFreeStyleProjectLabels(CauseData data, String expected)
```
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/environment/GitLabEnvironmentContributorTest.java]:
```suggestion
    public void testFreeStyleProjectLabels(CauseData causeData, String expected)
```

File: src/test/java/com/dabsquared/gitlabjenkins/environment/GitLabEnvironmentContributorTest.java [hunk 2]
```diff
@@ -74,46 +74,43 @@ public void matrixProjectTest() throws IOException, InterruptedException, Execut
...
             throws IOException, InterruptedException, ExecutionException {
         FreeStyleProject p = jenkins.createFreeStyleProject();
-        GitLabWebHookCause cause = new GitLabWebHookCause(causeData);
+        GitLabWebHookCause cause = new GitLabWebHookCause(data);
```
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/environment/GitLabEnvironmentContributorTest.java]:
```suggestion
        GitLabWebHookCause cause = new GitLabWebHookCause(causeData);
```

**Review by krisstern — CHANGES_REQUESTED** [review]

**Review by krisstern — CHANGES_REQUESTED** [review]
Hi @yiftahw, I think this implementation looks good but maybe we could add some instructions on how to use this feature before I can accept and merge this?

**Comment by krisstern** [comment]
You can basically just copy and past what you have in the description to the README we have at the root

**Comment by yiftahw** [comment]
> You can basically just copy and past what you have in the description to the README we have at the root

@krisstern thinking about what to put in the README, I thought to myself that people that will want to use this feature will probably suffer from some boilerplate of parsing the string held by the environment variable...  
I'm working now on adding a function called `GitLabMergeRequestLabelExists("some label") -> Boolean`
I'm kind of copying the mechanism used in `AcceptGitLabMergeRequestStep`
your thoughts?


example pipeline:
```groovy
script {
    if (GitLabMergeRequestLabelExists("bugfix"))
    {
        echo 'bugfix label detected!'
    }
}
```
instead of:  
```groovy
script {
    if (env.gitlabMergeRequestLabels) {
        def labels = env.gitlabMergeRequestLabels.split(',').collect { it.trim() }
        if ('bugfix' in labels) {
            echo 'bugfix label detected!'
        }
    }
}
```
I'm moving this to draft as I need to add some unit-tests but form initial testing with the docker setup it looks to be working fine

**Comment by krisstern** [comment]
Hi @yiftahw, regarding:

> I'm working now on adding a function called GitLabMergeRequestLabelExists("some label") -> Boolean
I'm kind of copying the mechanism used in AcceptGitLabMergeRequestStep
your thoughts

Sounds like a reasonable approach to me! Please go ahead with the implementation.

File: src/test/java/com/dabsquared/gitlabjenkins/workflow/GitLabMergeRequestLabelExistsStepTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * @author Yiftah Waisman
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class GitLabMergeRequestLabelExistsStepTest {
+
+    @ClassRule
+    public static JenkinsRule jenkins = new JenkinsRule();
+
+    private void test_webhook_base(CauseData causeData, String expected)
```
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/workflow/GitLabMergeRequestLabelExistsStepTest.java]:
```suggestion
    private void testWebhookBase(CauseData causeData, String expected)
```

File: src/test/java/com/dabsquared/gitlabjenkins/workflow/GitLabMergeRequestLabelExistsStepTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+        // wait for the build to finish and obtain the run object
+        WorkflowRun run = (WorkflowRun) future.get();
+
+        // observe the logs to see if the label was detected or not
+        jenkins.assertLogContains(expected, run);
+    }
+
+    @Test
+    public void test_label_exists_in_mr_webhook()
```
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/workflow/GitLabMergeRequestLabelExistsStepTest.java]:
```suggestion
    public void testLabelExistsInMrWebhook()
```

File: src/test/java/com/dabsquared/gitlabjenkins/workflow/GitLabMergeRequestLabelExistsStepTest.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            throws IOException, ExecutionException, InterruptedException, FormException {
+        // create a cause data object with a label
+        CauseData causeData = generateCauseDataWithLabels(Arrays.asList("test label"));
+        test_webhook_base(causeData, "test label found");
+    }
+
+    @Test
+    public void test_no_labels_in_mr_webhook()
```
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/workflow/GitLabMergeRequestLabelExistsStepTest.java]:
```suggestion
    public void testNoLabelsInMrWebhook()
```

**Review by krisstern — CHANGES_REQUESTED** [review]
Some more changes before we are good to go

**Comment by krisstern** [comment]
Please apply `mvn spotless:apply` to fix the `spotless` issues

**Review by krisstern — APPROVED** [review]
LGTM

**Comment by krisstern** [comment]
Hi @yiftahw the PR looks good now, thanks for your contribution! I will let it sit for between 24 and 48 hours before I merge it if there is no objection from the community.

**Comment by yiftahw** [comment]
@krisstern before merging, I have some changes that close #1715 , they are pretty minor, should I push these changes here and you'll review them?
[you can view them here](https://github.com/yiftahw/gitlab-plugin/commit/92fad7d07f1582cdf236ee356df468a6409ff8d2)  
TL;DR now both `MergeRequestAttributes` and `MergeRequestHook` have a member `List<MergeRequestLabel> labels`, which more accurately models the webhook payloads. (before that, only `MergeRequestHook` held the labels)

**Comment by krisstern** [comment]
Hi @yiftahw I think it may be better to create another PR to address issue #1715. I can review that and merge it before cutting a new release.
